colexec: disallow casts from decimal to unequal decimal

asubiotto opened the pull request:
Release justification: fixes decimal rounding errors in the
vectorized execution engine.

These could occur when, for example, rounding a decimal. The vectorized
execution engine does not support this cast because the conversion from types.T
to coltypes.T is lossy and therefore loses the rounding information. When
casting to a decimal, we now check that the decimal we're casting from is of
the same decimal type as the decimal type we're casting to and run this
expression through the row execution engine if not.

Release note (bug fix): bug fix to decimal rounding errors in the vectorized
execution engine

Fixes #45837

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/45950)
<!-- Reviewable:end -->

bobvawter commented:
[![CLA assistant check](https://cla.crdb.dev/pull/badge/not_signed)](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=45950) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=45950) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.crdb.dev/check/cockroachdb/cockroach?pullRequest=45950) it.</sub>

yuzefovich reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45950#-:-M24GbUP5mQuOS9Mn6xB:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

Commit:
colexec: disallow casts from decimal to unequal decimal

Release justification: fixes decimal rounding errors in the
vectorized execution engine.

These could occur when, for example, rounding a decimal. The vectorized
execution engine does not support this cast because the conversion from types.T
to coltypes.T is lossy and therefore loses the rounding information. When
casting to a decimal, we now check that the decimal we're casting from is of
the same decimal type as the decimal type we're casting to and run this
expression through the row execution engine if not.

Release note (bug fix): fixes decimal rounding errors in the vectorized
execution engine.

asubiotto commented:
bors r=yuzefovich

yuzefovich commented:
bors r-

It needs a release justification (probably the missing one is in the PR description, I'll add it).

yuzefovich commented:
bors r=yuzefovich
